File: lab3_16_12/1_1.py
import collections
import sys

def print_path(parent_src, parent_dest, m, src, dest):
    path_src = []
    curr = m
    while curr != -1:
        path_src.append(curr)
        if curr == src: break
        curr = parent_src[curr]
    path_src.reverse()

    path_dest = []
    curr = parent_dest[m]
    while curr != -1:
        path_dest.append(curr)
        if curr == dest: break
        curr = parent_dest[curr]
    
    full_path = path_src + path_dest
    print(*(full_path))

def bidirectional_bfs(adj, src, dest, n):
    if src == dest:
        print("src & dest are same.")
        return

    q_src = collections.deque([src])
    q_dest = collections.deque([dest])

    vis_src = [0] * (n + 1)
    vis_dest = [0] * (n + 1)
    parent_src = [-1] * (n + 1)
    parent_dest = [-1] * (n + 1)

    vis_src[src] = 1
    vis_dest[dest] = 1

    while q_src and q_dest:
        # From src
        if q_src:
            u = q_src.popleft()
            if u in adj:
                for v in adj[u]:
                    if not vis_src[v]:
                        vis_src[v] = 1
                        parent_src[v] = u
                        q_src.append(v)

                        if vis_dest[v]:
                            print_path(parent_src, parent_dest, v, src, dest)
                            return

        # From dest
        if q_dest:
            u = q_dest.popleft()
            if u in adj:
                for v in adj[u]:
                    if not vis_dest[v]:
                        vis_dest[v] = 1
                        parent_dest[v] = u
                        q_dest.append(v)

                        if vis_src[v]:
                            print_path(parent_src, parent_dest, v, src, dest)
                            return

    print("No path exists")

def main():
    try:
        input_data = sys.stdin.read().split()
        if not input_data: return
        it = iter(input_data)
        
        n = int(next(it))
        adj = collections.defaultdict(list)
        for i in range(n):
            for j in range(n):
                val = int(next(it))
                if val == 1:
                    adj[i].append(j)

        src = int(next(it))
        dest = int(next(it))

        bidirectional_bfs(adj, src, dest, n)
    except (StopIteration, ValueError, EOFError):
        pass

if __name__ == "__main__":
    main()

File: lab3_16_12/1_1.cpp
#include<bits/stdc++.h>

using namespace std;

void printPath(vector<int>& parent_src, vector<int>& parent_dest, int m, int src, int dest)
{
    vector<int> path;

    int curr = m;
    path.push_back(curr);

    while(curr != src)
    {
        curr = parent_src[curr];
        path.push_back(curr);
    }

    reverse(path.begin(), path.end());

    curr = m;
    while(curr != dest)
    {
        curr = parent_dest[curr];
        path.push_back(curr);
    }

    for(int i : path)
        cout << i << " ";
    cout << "\n";
}

void biBFS(vector<int> adj[], int src, int dest, int n)
{
    if(src == dest)
    {
        cout << "src & dest are same.\n";
        return;
    }
    queue<int> q_src, q_dest;

    vector<int> vis_src(n+1, 0);
    vector<int> vis_dest(n+1, 0);

    vector<int> parent_src(n+1, -1);
    vector<int> parent_dest(n+1, -1);

    q_src.push(src);
    q_dest.push(dest);

    vis_src[src] = 1;
    vis_dest[dest] = 1;

    while(!q_src.empty() && !q_dest.empty())
    {
        // From src
        if(!q_src.empty())
        {
            int u = q_src.front();
            q_src.pop();

            for(int v : adj[u])
            {
                if(!vis_src[v])
                {
                    vis_src[v] = 1;
                    parent_src[v] = u;
                    q_src.push(v);

                    if(vis_dest[v])
                    {
                        printPath(parent_src, parent_dest, v, src, dest);
                        return;
                    }
                }
            }
        }

        // From dest
        if(!q_dest.empty())
        {
            int u = q_dest.front();
            q_dest.pop();

            for(int v : adj[u])
            {
                if(!vis_dest[v])
                {
                    vis_dest[v] = 1;
                    parent_dest[v] = u;
                    q_dest.push(v);

                    if(vis_src[v])
                    {
                        printPath(parent_src, parent_dest, v, src, dest);
                        return;
                    }
                }
            }
        }
    }

    cout << "No path exists\n";
}

int main()
{
    int n;
    cout << "Enter no. of nodes : ";
    cin >> n;
    vector<int> adj[n+1];
    vector<vector<int>> grid(n, vector<int>(n));
    for(int i=0;i<n;i++)
    {
        for(int j= 0;j<n;j++)
        {
            cin >> grid[i][j];
            if(grid[i][j] == 1)
            {
                adj[i].push_back(j);
            }
        }
    }

    cout << "Enter src & dest : ";
    int src, dest;
    cin >> src >> dest;

    biBFS(adj, src, dest, n);

    return 0;
}